pkg/volume/csi/csi_attacher.go

```diff
@@ -335,10 +496,80 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
 }
 
 func (c *csiAttacher) UnmountDevice(deviceMountPath string) error {
-	glog.V(4).Info(log("detacher.UnmountDevice is not implemented"))
+	glog.V(4).Info(log("attacher.UnmountDevice(%s)", deviceMountPath))
+
+	// Setup
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	driverName, volID, err := getDriverAndVolNameFromDeviceMountPath(deviceMountPath, c.plugin)
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed to get driver and volume name from device mount path: %v", err))
+		return err
+	}
+
+	if c.csiClient == nil {
+		addr := fmt.Sprintf(csiAddrTemplate, driverName)
+		c.csiClient = newCsiDriverClient("unix", addr)
+	}
+	csi := c.csiClient
+
+	// Check whether "STAGE_UNSTAGE_VOLUME" is set
+	capabilities, err := csi.NodeGetCapabilities(ctx)
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed while getting node capabilities: %v", err))
+		return err
+	}
+	if capabilities == nil {
+		glog.Infof(log("attacher.UnmountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping UnountDevice..."))
+		return nil
+	}
+
+	stageUnstageSet, err := hasStageUnstageCapability(capabilities)
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed: %v", err))
+		return err
+	}
+	if !stageUnstageSet {
+		glog.Infof(log("attacher.UnmountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping UnmountDevice..."))
+		return nil
+	}
+
+	// Start UnmountDevice
+	err = csi.NodeUnstageVolume(ctx,
+		volID,
+		deviceMountPath)
+
+	if err != nil {
+		glog.Errorf(log("attacher.UnmountDevice failed: %v", err))
+		return err
+	}
+
+	glog.V(4).Infof(log("attacher.UnmountDevice successfully requested NodeStageVolume [%s]", deviceMountPath))
 	return nil
 }
 
+func hasStageUnstageCapability(capabilities []*csipb.NodeServiceCapability) (bool, error) {
+	stageUnstageSet := false
+	for _, capability := range capabilities {
+		if capability.GetRpc().GetType() == csipb.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME {
+			stageUnstageSet = true
+		}
+	}
+	return stageUnstageSet, nil
+}
+
+func getDriverAndVolNameFromDeviceMountPath(deviceMountPath string, plugin *csiPlugin) (string, string, error) {
```

Comment:
Oh man this is a fun bit of code.

Reply:
It would be helpful to document the format deviceMountPath (i.e. `{pluginDir}/csi/etc` so future maintainers would have an idea why this works.